tikv/raftstore/fsm_apply.go

```diff
@@ -1037,34 +1053,300 @@ func (d *applyDelegate) createWriteCmdOps(requests []*raft_cmdpb.Request) (ops w
+				d.tag, peer, d.region)
+			log.Error(errMsg)
+			err = errors.New(errMsg)
+			return
+		}
+		region.Peers = append(region.Peers, peer)
+		log.Infof("%s add learner successfully, peer %s, region %s", d.tag, peer, d.region)
+	}
+	state := rspb.PeerState_Normal
+	if d.pendingRemove {
+		state = rspb.PeerState_Tombstone
+	}
+	WritePeerState(aCtx.wb, region, state, nil)
+	resp = &raft_cmdpb.AdminResponse{
+		ChangePeer: &raft_cmdpb.ChangePeerResponse{
+			Region: region,
+		},
+	}
+	result = applyResult{
+		tp: applyResultTypeExecResult,
+		data: &execResultChangePeer{
+			cp: changePeer{
+				confChange: new(eraftpb.ConfChange),
+				region:     region,
+				peer:       peer,
+			},
+		},
+	}
+	return
 }
 
 func (d *applyDelegate) execSplit(aCtx *applyContext, req *raft_cmdpb.AdminRequest) (
 	resp *raft_cmdpb.AdminResponse, result applyResult, err error) {
-	return // TODO: stub
+	split := req.Split
+	adminReq := &raft_cmdpb.AdminRequest{
+		Splits: &raft_cmdpb.BatchSplitRequest{
+			Requests:    []*raft_cmdpb.SplitRequest{split},
+			RightDerive: split.RightDerive,
+		},
+	}
+	// This method is executed only when there are unapplied entries after being restarted.
+	// So there will be no callback, it's OK to return a response that does not matched
+	// with its request.
+	return d.execBatchSplit(aCtx, adminReq)
 }
 
 func (d *applyDelegate) execBatchSplit(aCtx *applyContext, req *raft_cmdpb.AdminRequest) (
 	resp *raft_cmdpb.AdminResponse, result applyResult, err error) {
-	return // TODO: stub
+	splitReqs := req.Splits
+	rightDerive := splitReqs.RightDerive
+	if len(splitReqs.Requests) == 0 {
+		err = errors.New("missing split key")
+		return
+	}
+	derived := new(metapb.Region)
+	if err := CloneMsg(d.region, derived); err != nil {
+		panic(err)
+	}
+	newRegionCnt := len(splitReqs.Requests)
+	regions := make([]*metapb.Region, 0, newRegionCnt+1)
+	keys := make([][]byte, 0, newRegionCnt+1)
+	keys = append(keys, derived.StartKey)
+	for _, request := range splitReqs.Requests {
+		splitKey := request.SplitKey
+		if len(splitKey) == 0 {
+			err = errors.New("missing split key")
+			return
+		}
+		if bytes.Compare(splitKey, keys[len(keys)-1]) <= 0 {
+			err = errors.Errorf("invalid split request:%s", splitReqs)
+			return
+		}
+		if len(request.NewPeerIds) != len(derived.Peers) {
+			err = errors.Errorf("invalid new peer id count, need %d but got %d",
+				len(derived.Peers), len(request.NewPeerIds))
+			return
+		}
+		keys = append(keys, splitKey)
+	}
+	keys = append(keys, derived.EndKey)
+	err = CheckKeyInRegion(keys[len(keys)-2], d.region)
+	if err != nil {
+		return
+	}
+	log.Infof("%s split region %s, keys %v", d.tag, d.region, keys)
+	derived.RegionEpoch.Version += uint64(newRegionCnt)
+	// Note that the split requests only contain ids for new regions, so we need
+	// to handle new regions and old region separately.
+	if !rightDerive {
+		derived.EndKey = keys[1]
+		keys = keys[1:]
+		regions = append(regions, derived)
+	}
```

Comment:
Else you should drop the last key in keys?

Reply:
No need, the last key will not be used.